ublox-bluetooth-commands.go

```diff
@@ -171,12 +171,23 @@ func (ub *UbloxBluetooth) discoveryCommandWithContext(ctx context.Context, scant
 // ConnectToDevice attempts to connect to the device with the specified address.
 func (ub *UbloxBluetooth) ConnectToDevice(address string, onConnect DeviceEvent, onDisconnect DeviceEvent) (err error) {
 	var d []byte
+	var cr *ConnectionReply
+
 	d, err = ub.writeAndWait(ConnectCommand(address), true)
 	if err != nil {
 		return err
 	}
 
-	var cr *ConnectionReply
+	// Flush out disconnect request - Bug in current Nina firmware.
+	err = ub.ATCommand()
+	if err != nil {
+		// Try again...
+		d, err = ub.writeAndWait(ConnectCommand(address), true)
```

Comment:
Presumably this is a single retry in the event that we encounter this NINA bug.

It is, of course, possible that the CONNECT request fails to reach the sensor (RF issues, etc) and so these retries are handled elsewhere.

I don't think that this code is quite right though: The connect request will result in receving a +UUBTACLC packet from the NINA. The bug is that the NINA immediately sends a +UUBTACLD as well but this is blocked until flushed out by another AT command.

What you really want to do is send the "AT" command and this result will be OK however you need to see if you have also received the (asynchronous) +UUBTACLD response and are no longer in a connection.

I'm not clear how the Go code handles connections

Reply:
Looking at the code, there is an onDisconnect event callback mechanism defined for connectToDevice() calls but it is not connected up. The only code I can see that looks for the Disconnect reply message is in parseEDMMessage() which parses the raw EDM replies, but all it does is print ""Unexpected Disconnect" and it does not return an error.

Reply:
The library has always been written with the understanding that the retry logic is performed at the application layer.
In the Unexpected Disconnect handler I would then create an Error constant ErrorUnexpectedDisconnect, which can be thrown at any time.
In the generic retry handler I would then put the procedure to 'belch' the disconnect, and then rety.